ncm-iptables/src/main/perl/iptables.pm

```diff
@@ -18,311 +13,332 @@ use vars qw(@ISA $EC);
+    '--out-interface'    => '\!?\s*\w+',
+    '--match'            => '(tcp|udp|igmp|all|icmp|state|limit|owner|mark|ttl|pkttype|recent|unclean|length|multiport|conntrack)',
+    '--pkt-type'         => '(unicast|broadcast|multicast)',
+    '--mark'             => '(\d+|\d+/\d+)',
+    '--ttl'              => '\d+',
+    '--tos'              => '\S+',
+    '--sid-owner'        => '\d+',
+    '--state'            => '(\!?\s+|)(new|established|related|invalid)',
+    '--limit'            => '\S+',
+    '--syn'              => '',
+    '! --syn'            => '',
+    '--icmp-type'        => '\w+',
+    '-j'                 => "",
+    '--log-prefix'       => '\w+',
+    '--log-level'        => '(debug|info|notice|warning|warn|err|error|crit|alert|emerg|panic)',
+    '--log-tcp-sequence' => '',
+    '--log-tcp-options'  => '',
+    '--log-ip-options'   => '',
+    '--log-uid'          => '',
+    '--reject-with'      => '(icmp-net-unreachable|icmp-host-unreachable|icmp-port-unreachable|icmp-proto-unreachable|icmp-net-prohibited|icmp-host-prohibited|tcp-reset)',
+    '--set-class'        => '\d{1,2}:\d{1,2}',
+    '--limit-burst'      => '\S+',
+    '--to-destination'   => '\S+',
+    '--to-source'        => '\S+',
+    '--to-ports'         => '\d+(-\d+)?',
+    '--uid-owner'        => '\d+',
+    '--tcp-flags'        => '\S+',
+    '--tcp-option'       => '\d+',
+    '--length'           => '(\d+|\d+:\d+)',
+    '--ctstate'          => '(new|established|related|invalid|snat|dnat)(,(new|established|related|invalid|snat|dnat))*',
+    '--fragment'         => '',
+    '! --fragment'       => '',
+    '--set'              => '',
+    '--rcheck'           => '',
+    '--update'           => '',
+    '--remove'           => '',
+    '--rdest'            => '',
+    '--rsource'          => '',
+    '--rttl'             => '',
+    '--seconds'          => '\d+',
+    '--hitcount'         => '\d+',
+    '--name'             => '\S+',
+    '--comment'          => '\S+',
+);
 
 # Operations to perform on the resource options when read for the first time.
-my %options_op  = ( '-A'              => \&upercase,
-                    '-D'              => \&upercase,
-                    '-I'              => \&upercase,
-                    '-R'              => \&upercase,
-                    '-N'              => \&upercase,
-                    '-j'              => \&upercase,
-                    '-s'              => \&dns2ip,
-                    '-d'              => \&dns2ip
-    );
+Readonly::Hash my %OPTION_MODIFIERS => (
+    '-A' => 'uppercase',
+    '-D' => 'uppercase',
+    '-I' => 'uppercase',
+    '-R' => 'uppercase',
+    '-N' => 'uppercase',
+    '-j' => 'uppercase',
+    '-s' => 'dns2ip',
+    '-d' => 'dns2ip',
+    '--comment' => 'quote_string',
+);
+
+sub trim_whitespace {
+    # Trim leading and trailing whitespace
+    my ($self, $text) = @_;
+    $text =~ s/^\s+|\s+$//g;
+    return defined $text ? $text : '';
```

Comment:
if `$text` is not defined, line above will produce a warning.
```perl
sub trim_whitespace
{
    my ($self, $text) = @_;
    $text = '' if ! defined $text;
    $text =~ s/^\s+|\s+$//g;
    return $text;
}
```
perl 5.10 has a defined-or operator, but not in el5...